File: pa1/List.java
/*
Student: Nixon Duong
CruzID: niduong
Professor: Patrick Tantalo
Course: CMPS101
Programming Assignment: PA1
Description: Provides a bi-directional list to clients
File: List.java
Clients: Lex.java
*/
public class List{
	private class Node{
		int data;
		Node prev = null;
		Node next = null;
		Node(int data){
			this.data = data;
		}
		public String toString(){
			return Integer.toString(data);
		}
	}
	private Node front,back,cursor;
	private int length;
	private int index;
	
	//Constructor
	List(){// Creates a new empty list.
		front = back = cursor = null;
		length = 0;
		index = -1;
	}

	//Access functions
	int length(){//Returns the number of elements in this List.
		return this.length;
	}

	int index(){//If cursor is defined, returns the index of the cursor element, otherwise -1.
		if(this.cursor != null){
			return this.index;
		}
		return -1;
	}

	int front(){//Returns front element. Pre: length()>0
		if(this.length>0){
			return this.front.data;
		}
		return -1;
	}

	int back(){//Returns back element. Pre: length()>0
		if(this.length>0){
			return this.back.data;
		}
		return -1;
	}

    int get(){//Returns cursor element. Pre: length()>0, index() >=0
	if(length()>0){
	    if(cursor != null){
			return cursor.data;
	    }
	    else{
			return -1;
	    }
	}
	return -1;
    }
	    
	boolean equals(List L){//Returns true if and only if this List and L are the same integer sequence. The states of the cursors in the two List are not used in determining equality
		return(this.toString().equals(L.toString()));
	}

	//Manipulation procedures
	void clear(){//Resets this List to its original empty state
		this.front = null;
		this.back = null;
		this.cursor = null;
		this.length = 0;
		this.index = -1;
	}

	void moveFront(){//If List is non-empty, places cursor under the front element, otherwise does nothing
		if(this.length>0){
			this.cursor = this.front;
		}
		this.index = 0;
	}

	void moveBack(){//If List is non-empty, places the cursor under the back element, otherwise does nothing
		if(this.length>0){
			this.cursor = this.back;
			this.index = this.length - 1;
		}
	}

	void movePrev(){//If cursor is defined and not at front, moves cursor one step toward front of this List, if cursor is defined and at the front, cursor becomes undefined, if cursor is undefined does nothing
		if(this.cursor!=null){
			if(this.cursor!=this.front){
				this.cursor = this.cursor.prev;
			}
			else{
				this.cursor = null;
			}
		}
		this.index--;
	}

	void moveNext(){//If cursor is defined and not at back, moves cursor one step toward back of this List, if cursor is defined and at the back, cursor becomes undefined, if cursor is undefined does nothing
		if(this.cursor!=null){
			if(this.cursor!=this.back){
				this.cursor = this.cursor.next;
			}
			else{
				this.cursor = null;
			}
		}
		this.index++;
	}

	void prepend(int data){//Insert new element into this List. If List is non-empty, insertion takes place before front element.
		Node newNode = new Node(data);
		if(this.length == 0){
			this.front = this.back = newNode;
		}
		else{
			newNode.next = this.front;
			this.front.prev = newNode;
			this.front = newNode;
		}
		this.length++;
		this.index++;
	}

	void append(int data){//Insert new element into this List. If List is non-empty, insertion takes place after back element.
		Node newNode = new Node(data);
		if(this.length == 0){
			this.front = this.back = newNode;
		}
		else{
			this.back.next = newNode;
			newNode.prev = this.back;
			this.back = newNode;
		}
		length++;
	}

	void insertBefore(int data){//Insert new element before cursor. Pre: length()>0, index()>=0
		if(this.length > 0 && this.index >=0){
			Node newNode = new Node(data);
			if(this.cursor == this.front){
				this.prepend(data);
			}
			else{
				newNode.next = this.cursor;
				this.cursor.prev.next = newNode;
				newNode.prev = this.cursor.prev;
				this.cursor.prev = newNode;
				this.length++;
				this.index++;
			}
		}
	}

	void insertAfter(int data){//Insert new element after cursor. Pre: length()>0, index()>=0
		if(this.length>0 && this.index >=0){
			Node newNode = new Node(data);
			if(this.cursor == this.back){
				this.append(data);
			}
			else{
				newNode.prev = this.cursor;
				newNode.next = this.cursor.next;
				this.cursor.next.prev = newNode;
				this.cursor = newNode;
				this.length++;
			}
		}
	}

	void deleteFront(){//Deletes the front element. Pre: length()>0
		if(this.length == 0){
			throw new RuntimeException("deleteFront() on emptyList");
		}
		if(this.length > 0){
			if(this.length == 1){
				this.cursor = this.front = this.back = null;
				this.index = -1;
			}
			else{
				if(this.cursor == this.front){
					this.front = this.front.next;
					this.front.prev = null;
					this.cursor = null;
					this.index = -1;
				}
				else{
					this.front = this.front.next;
					this.front.prev = null;
					if(this.cursor!=null){
						this.index--;
					}
				}
			}
			this.length--;
		}
	}

	void deleteBack(){//Deletes the back element. Pre: length()>0
		if(this.length == 0){
			throw new RuntimeException("deleteBack() on emptyList");
		}
		if(this.length > 0){
			if(this.length == 1){
				this.cursor = this.front = this.back = null;
				this.index = -1;
			}
		else{
			if(this.cursor == this.back){
				this.cursor = null;
				this.index = -1;
			}
			this.back = this.back.prev;
			this.back.next = null;
		}
			this.length--;
		}
	}

	void delete(){//Deletes cursor element, making cursor undefined. Pre: length()>0, index()>=0
		if(this.length == 0){
			throw new RuntimeException("delete() on emptyList");
		}
		if(this.length>0 && this.index >=0){
			if(this.cursor == this.front){
				this.deleteFront();
			}
			else if(this.cursor == this.back){
				this.deleteBack();
			}
			else{
				this.cursor.prev.next = this.cursor.next;
				this.cursor.next.prev = this.cursor.prev;
				this.length--;
			}
			this.cursor = null;
			this.index = -1;
		}
	}

	//Other methods
	public String toString(){//Overrides Object's toStrig method. Returns a String representation of this List consistign of a space separated sequence of integers, with front on left
		if(this.length == 0){
			return "";
		}
		else{
			Node iterator = this.front;
			StringBuffer list = new StringBuffer();
			while(iterator != null){
				list.append(iterator.toString()).append(" ");
				iterator = iterator.next;
			}
			return (new String(list));
		}
	}

	List copy(){//Returns a new List representing the same integer sequence as this List. The cursor in this new list in undefined, regardless of state of the cursor in this list. This list is unchanged.
		List newList = new List();
		if(this.length!=0){
			Node iterator = this.front;
			while(iterator!=null){
				newList.append(iterator.data);
				iterator = iterator.next;
			}
			return newList;
		}
		return newList;
	}
}

File: pa4/List.c
/*
Student: Nixon Duong
CruzID: niduong
Professor: Patrick Tantalo
Course: CMPS101
Programming Assignment: PA4
Description: Provides a bi-directional list to clients
File: List.c
Clients: Graph.c
*/
#include <stdio.h>
#include <stdlib.h>
#include "List.h"

// structs
typedef struct NodeObj{
	int data;
	struct NodeObj* next;
	struct NodeObj* prev;
} NodeObj;

typedef NodeObj* Node;

typedef struct ListObj{
	Node front;
	Node back;
	Node cursor;
	int length;
	int index;
} ListObj;  

Node newNode(int data){
	Node node = malloc(sizeof(NodeObj));
	node->data = data;
	node->next = NULL;
	node->prev = NULL;
	return(node);
}

void freeNode(Node *pN){
	if(pN!=NULL && *pN!=NULL){
		free(*pN);
		*pN = NULL;
	}
}

List newList(void){
	List L;
	L = malloc(sizeof(ListObj));
	L->front = L->back = L->cursor = NULL;
	L->length = 0;
	L->index = -1;
	return(L);   
}

//Access functions
int length(List L){// returns the number of elements in the List
	if(L==NULL){
		printf("List Error: called length() on NULL List reference\n");
		exit(1);
	}
	else{
		return(L->length);
	}
}


void freeList(List *pL){
	if(pL!=NULL && *pL!=NULL){
		while(length(*pL) > 0){
			deleteFront(*pL);
		}
		free(*pL);
		*pL = NULL;
	}
}


int index(List L){// if cursor defined, returns index of cursor, else -1
	if(L==NULL){
		printf("List Error: called index() on NULL List reference\n");
		exit(1);
	}
	else{
		return(L->index); // if cursor isnt defined, the index is -1
	}
}

int front(List L){// returns the front element Pre: len > 0
	if(L==NULL){
		printf("List Error: called front() on NULL List reference\n");
		exit(1);
	}
	else{
		if(L->length >0){
			return(L->front->data);
		}
		return(-1);
	}
}

int back(List L){// return the back element Pre: len > 0
	if(L==NULL){
		printf("List Error: called back() on NULL List reference\n");
		exit(1);
	}	
	else{
		if(L->length > 0){
			return(L->back->data);
		}
		return(-1);
	}
}

int get(List L){ // Returns cursor element Pre: len > 0 , index >=0
	if(L==NULL){
		printf("List Error: called get() on NULL List reference\n");
		exit(1);
	}
	else{
		if(L->length>0 && L->index>=0){
			return(L->cursor->data);
		}
		return(-1);
	}
}

int equals(List A, List B){// Returns 1 if List A is equal to List B, else 0
	int equal = 0;
	Node X = NULL;
	Node Y = NULL;
	if(A==NULL || B==NULL){
		printf("List Error: called equals() on NULL List reference\n");
		exit(-1);
	}
	equal = (A->length == B->length);
	X = A->front;
	Y = B->front;
	while(equal && X!=NULL){
		equal = (X->data == Y->data);
		X = X->next;
		Y = Y->next;
	}
	return(equal);
}

//Manipulation functions
void clear(List L){// resets the list
	if(L==NULL){
		printf("List Error: called clear() on NULL List reference\n");
		exit(-1);
	}
	else{
		while(length(L) > 0){
			deleteFront(L);
		}
		L->front = L->back = L->cursor = NULL;
		L->length = 0;
		L->index = -1;
	}
}

void moveFront(List L){// if list non-empty, place cursor on front node, else does nothing
	if(L==NULL){
		printf("List Error: called moveFront() on NULL List reference\n");
		exit(-1);
	}
	else{
		if(L->length > 0){
			L->cursor = L->front;
			L->index = 0;
		}
	}
}

void moveBack(List L){// if list non-empty, place cursor on back node, else does nothing
	if(L==NULL){
		printf("List Error: called moveBack() on NULL List reference\n");
		exit(-1);
	}
	else{
		if(L->length >0){
			L->cursor = L->back;
			L->index = L->length - 1;
		}
	}
}

void movePrev(List L){// if cursor is defined and not front, move cursor prev, if defined and at front, cursor is undefined
	if(L==NULL){
		printf("List Error: called movePrev() on NULL List reference\n");
		exit(-1);
	}
	if(L->cursor!=NULL){
		if(L->cursor != L->front){
			L->cursor = L->cursor->prev;
		}
		else{
			L->cursor = NULL;
		}
		L->index--;
	}
	
}
void moveNext(List L){// if cursor is defined and not back, move cursor back, if defined and at back, cursor is undefined
	if(L==NULL){
		printf("List Error: called moveNext() on NULL List reference\n");
		exit(-1);
	}
	if(L->cursor!=NULL){
		if(L->cursor != L->back){
			L->cursor = L->cursor->next;
			L->index++;
		}
		else{
			L->cursor = NULL;
			L->index = -1;
		}
	}
}

void prepend(List L, int data){// insert new element into the list, if list is non-empty, element goes in the front
	if(L==NULL){
		printf("List Error: called prepend() on NULL List reference\n");
		exit(-1);
	}
	else{
		Node node = newNode(data);
		if(L->length == 0){
			L->front = L->back = node;
		}
		else{
			node->next = L->front;
			L->front->prev = node;
			L->front = node;
		}
		L->index++;
		L->length++;
	}
}

void append(List L, int data){// insert new element into the list, if list is non-empty, element goes in the back
	if(L==NULL){
		printf("List Error: called append() on NULL List reference\n");
		exit(-1);
	}
	else{
		Node node = newNode(data);
		if(L->length == 0){
			L->front = L->back = node;
		}
		else{
			node->prev = L->back;
			L->back->next = node;
			L->back = node;
		}
		L->length++;
	}
}

void insertBefore(List L, int data){// insert new element before cursor. Pre: len > 0, index >=0
	if(L==NULL){
		printf("List Error: called insertBefore() on NULL List reference\n");
		exit(-1);
	}
	else{
		if(L->length > 0 && L->index >=0){
			if(L->cursor == L->front){
				prepend(L,data);
			}
			else{
				Node node = newNode(data);
				node->next = L->cursor;
				L->cursor->prev->next = node;
				node->prev = L->cursor->prev;
				L->cursor->prev = node;
				L->length++;
				L->index++;
			}
		}
	}
}

void insertAfter(List L, int data){// insert new element after cursor. Pre: len > 0, index >=0
	if(L==NULL){
		printf("List Error: called insertAfter() on NULL List reference\n");
		exit(-1);
	}
	else{
		if(L->length > 0 && L->index >=0){
			if(L->cursor == L->back){
				append(L,data);
			}
			else{
				Node node = newNode(data);
				node->prev = L->cursor;
				node->next = L->cursor->next;
				L->cursor->next->prev = node;
				L->cursor->next = node;
				L->length++;
			}
		}
	}
}

void deleteFront(List L){// deletes the front element. Pre: len > 0
	if(L==NULL){
		printf("List Error: called deleteFront() on NULL List reference\n");
		exit(-1);
	}
	else{
		if(L->length == 0){
			printf("deleteFront() on emptyList");
			exit(-1);
		}
		if(L->length > 0){
			if(L->length == 1){
				freeNode(&L->front);
				L->cursor = L->front = L->back = NULL;
				L->index = -1;
			}
			else{
				Node temp = NULL;
				if(L->cursor == L->front){
					temp = L->front;
					L->front = L->front->next;
					L->front->prev = NULL;
					L->cursor = NULL;
					L->index = -1;
				}
				else{
					temp = L->front;
					L->front = L->front->next;
					L->front->prev = NULL;
					if(L->cursor!=NULL){
						L->index--;
					}
				}
				freeNode(&temp);
			}
			L->length--;
		}
	}
}

void deleteBack(List L){// deletes the back element. Pre: len > 0
	if(L==NULL){
		printf("List Error: called deleteBack() on NULL List reference\n");
		exit(-1);
	}
	else{
		if(L->length == 0){
			printf("deleteBack() on emptyList");
			exit(-1);
		}
		if(L->length > 0){
			Node temp = NULL;
			if(L->length == 1){
				temp = L->back;
				L->cursor = L->front = L->back = NULL;
				L->index = -1;
			}
			else{
				if(L->cursor == L->back){
					L->cursor = NULL;
					L->index = -1;
				}
				temp = L->back;
				L->back = L->back->prev;
				L->back->next = NULL;
			}
			L->length--;
			freeNode(&temp);
		}
	}
}

void delete(List L){// deletes the cursor element, sets cursor to undefeined and index to -1 Pre: length > 0, index >=0
	if(L==NULL){
		printf("List Error: called delete() on NULL List reference\n");
		exit(-1);
	}
	else{
		if(L->length == 0){
			printf("delete() on emptyList");
			exit(-1);
		}
		if(L->length>0 && L->index >=0){
			if(L->cursor == L->front){
				deleteFront(L);
			}
			else if(L->cursor == L->back){
				deleteBack(L);
			}
			else{
				L->cursor->prev->next = L->cursor->next;
				L->cursor->next->prev = L->cursor->prev;
				L->length--;
			}
			freeNode(&L->cursor);
			L->index = -1;
		}
	}
}

//Other operations
void printList(FILE* out, List L){// prints list to file pointed by out, formatted as a space-separed string
	if(L==NULL){
		printf("List Error: called printList() on NULL List reference\n");
		exit(-1);
	}
	else{
		Node iterator = L->front;
		while(iterator!=NULL){
			fprintf(out,"%d ",iterator->data);
			iterator = iterator->next;
		}
	}
}
List copyList(List L){
	List list = newList();
	if(L==NULL){
		printf("List Error: called copyList() on NULL List reference\n");
		exit(-1);
	}
	else{
		if(L->length!=0){
			Node iterator = L->front;
			while(iterator!=NULL){
				append(list,iterator->data);
				iterator = iterator->next;
			}
			return list;
		}
	}
	return list;
}
